src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp

```diff
@@ -2,6 +2,69 @@
 
 
 extern double CCSN_MaxRefine_RadFac;
+extern double CCSN_CentralDens;
+
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Flag_CoreCollapse
+// Description :  Check if the element (i,j,k) of the input data satisfies the user-defined flag criteria
+//
+// Note        :  1. Invoked by "Flag_Check" using the function pointer "Flag_User_Ptr"
+//                   --> The function pointer may be reset by various test problem initializers, in which case
+//                       this function will become useless
+//                2. Enabled by the runtime option "OPT__FLAG_USER"
+//
+// Parameter   :  i,j,k       : Indices of the target element in the patch ptr[ amr->FluSg[lv] ][lv][PID]
+//                lv          : Refinement level of the target patch
+//                PID         : ID of the target patch
+//                Threshold   : User-provided threshold for the flag operation, which is loaded from the
+//                              file "Input__Flag_User"
+//                              In order of radius_min, radius_max, threshold_dens
+//
+// Return      :  "true"  if the flag criteria are satisfied
+//                "false" if the flag criteria are not satisfied
+//-------------------------------------------------------------------------------------------------------
+bool Flag_CoreCollapse( const int i, const int j, const int k, const int lv, const int PID, const double *Threshold )
+{
+
+   bool Flag = false;
+   bool MaxRefine = false;
+
+   const double dh        = amr->dh[lv];
+   const double Center[3] = { amr->BoxCenter[0], amr->BoxCenter[1], amr->BoxCenter[2] };
+   const double Pos   [3] = { amr->patch[0][lv][PID]->EdgeL[0] + (i+0.5)*dh,
+                              amr->patch[0][lv][PID]->EdgeL[1] + (j+0.5)*dh,
+                              amr->patch[0][lv][PID]->EdgeL[2] + (k+0.5)*dh  };
+
+   const double dx = Center[0] - Pos[0];
+   const double dy = Center[1] - Pos[1];
+   const double dz = Center[2] - Pos[2];
+   const double r  = sqrt(  SQR( dx ) + SQR( dy ) + SQR( dz )  );
+
+// (1) check if the allowed maximum level is reached
+   if ( CCSN_CentralDens < 1e11 )
+   {
+      MaxRefine = lv >= ( MAX_LEVEL - 2 );
+   }
+
+   else if ( CCSN_CentralDens < 1e12 )
+   {
+      MaxRefine = lv >= ( MAX_LEVEL - 1 );
+   }
+
+
+// (2) always refined to highest level in the region with r < 30 km, if allowed
+   if (  !MaxRefine  &&  ( r * UNIT_L < 3e6 )  )
+      Flag = true;
+
+//   if ( r * UNIT_L < 3e6 )
+//   printf("lv = %d, r = %.4e, CCSN_CentralDens = %.4e, MaxRefine = %d, MAX_LEVEL = %d, Flag = %d\n",
+//          lv, r * UNIT_L, CCSN_CentralDens, MaxRefine, MAX_LEVEL, Flag );
```

Comment:
If you want to retain this check, re-write it more formally.

Reply:
I removed this part.